parlai/core/context.py

```diff
@@ -0,0 +1,118 @@
+#!/usr/bin/env python3
+
+# Copyright (c) 2017-present, Facebook, Inc.
+# All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree. An additional grant
+# of patent rights can be found in the PATENTS file in the same directory.
+
+
+class BaseContext():
+    """Any XContext should implement the following set of functions. It can
+    then choose to implement any utility functions that it believes a model
+    especially tuned to use that context type will find useful.
+    """
+    def __init__(self, args):
+        """This function is used to directly create a context object"""
+        raise NotImplementedError
+
+    @staticmethod
+    def from_data(input_data):
+        """This function should take an input data object and produce the
+        context representation for this data. It is expected to be used by
+        tasks that are constructing context, and should be the inverse of
+        to_data_form
+        """
+        raise NotImplementedError
+
+    def to_data_form(self):
+        """This function should return an object that is a less structured
+        representation of the data contained in this object (generally in
+        the form of a dict or an array)
+        """
+        raise NotImplementedError
+
+    def to_context_string(self):
+        """This function should return a string that can be prepended to an
+        act's text in order to support models that don't use context objects
+        """
+        raise NotImplementedError
+
+    def to_display_string(self):
+        """This function will be invoked by the context binder whenever an
+        act with a context binder is printed
+        """
+        return self.to_data_form()
+
+    def get_type(self):
```

Comment:
Why wouldn't we want to use something like `isinstance` instead?

Reply:
Initially I had concerns about using Instance names because they aren't directly serializable. I can still reflect on this though